### PR TITLE
STM32 RTC sub seconds support

### PR DESCRIPTION
**Pull request by niedzwiecki-dawid**
Implement support for handling the sub second registers.

### CODE REVIEW

**Review by gautierg-st — CHANGES_REQUESTED** [review]
Some comments on new magic values that could be replaced. Otherwise LGTM

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 1]
```diff
@@ -61,6 +61,28 @@ LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
 #define COUNTER_NO_DATE
 #endif
 
+#if DT_INST_CLOCKS_CELL(1, bus) == STM32_SRC_LSI
+#if !defined(CONFIG_SOC_SERIES_STM32F1X)
+/* LSI @ 32 KHz */
+#define RTCCLK_FREQ 32000
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Make sense :-) done.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 2]
```diff
@@ -61,6 +61,28 @@ LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
...
+#else /* CONFIG_SOC_SERIES_STM32F1X */
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Done.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 3]
```diff
@@ -61,6 +61,28 @@ LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
...
+/* LSI @ 40 KHz */
+#define RTCCLK_FREQ 40000
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+#else
+/* LSE @ 32768 Hz */
+#define RTCCLK_FREQ 32768
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Done.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 4]
```diff
@@ -61,6 +61,28 @@ LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
...
+#endif /* DT_INST_CLOCKS_CELL(1, bus) == STM32_SRC_LSI */
+
+#if !defined(CONFIG_SOC_SERIES_STM32F1X)
+#define RTC_ASYNCPRE 127
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Done.

**Comment by semihalf-niedzwiecki-dawid** [comment]
I haven't payed attention to that earlier, but the clock source was checked incorrectly. I've fixed that in the new, first commit.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 5]
```diff
@@ -181,12 +186,25 @@ static int rtc_stm32_stop(const struct device *dev)
 }
 
 
-static uint32_t rtc_stm32_read(const struct device *dev)
+#ifndef CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED
+static uint32_t
+#else
+static uint64_t
+#endif
```
Comment (gautierg-st) [inline drivers/counter/counter_ll_stm32_rtc.c:193]: I suggest you define a 32 or 64-bit type once at the top depending on the subsecond config, and then use this type each time you need it instead of having an #Ifdef/#else/#endif each time. That will reduce the number of different cases and ease future maintenance.

**Review by gautierg-st — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 6]
```diff
@@ -164,20 +186,50 @@ static int rtc_stm32_stop(const struct device *dev)
...
+	do {
+		rtc_date = LL_RTC_DATE_Get(RTC);
+
+		do {
+			rtc_time = LL_RTC_TIME_Get(RTC);
+
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED
+			do {
+				rtc_subseconds = LL_RTC_TIME_GetSubSecond(RTC);
+			} while (rtc_subseconds != LL_RTC_TIME_GetSubSecond(RTC));
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: You are right. Reading the last register two times doesn't make sense.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 7]
```diff
@@ -228,7 +253,15 @@ static uint32_t rtc_stm32_read(const struct device *dev)
 	ts -= T_TIME_OFFSET;
 
 	__ASSERT(sizeof(time_t) == 8, "unexpected time_t definition");
+
+
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED
+	ticks = ts * counter_get_frequency(dev);
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c:252]: True. Done.

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 5 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Very good suggestion, thanks. Done.

**Review by gautierg-st — DISMISSED** [review]
LGTM! Thanks

**Review by erwango — CHANGES_REQUESTED** [review]
Please add some (at least minimal) way to get this compiled in CI

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 8]
```diff
@@ -164,20 +192,39 @@ static int rtc_stm32_stop(const struct device *dev)
...
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED
+		do {
+			rtc_time = LL_RTC_TIME_Get(RTC);
+			rtc_subseconds = LL_RTC_TIME_GetSubSecond(RTC);
+		} while (rtc_time != LL_RTC_TIME_Get(RTC));
+#else /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS_BASED */
+		rtc_time = LL_RTC_TIME_Get(RTC);
+#endif
+
+	} while (rtc_date != LL_RTC_DATE_Get(RTC));
```
Comment (faloj) [inline drivers/counter/counter_ll_stm32_rtc.c:226]: Reference manual says : "To ensure consistency between the 3 values, reading either RTC_SSR or RTC_TR locks the values in the higher-order calendar shadow registers until RTC_DR is read".
So imho you must call LL_RTC_TIME_GetSubSecond() before LL_RTC_TIME_Get() and LL_RTC_DATE_Get().

**Comment by semihalf-niedzwiecki-dawid** [comment]
Sorry for the delay, I'm working on the comments now.

**Comment by semihalf-niedzwiecki-dawid** [comment]
> Please add some (at least minimal) way to get this compiled in CI

Do you have any suggestions how to do it?

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 8 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c:226]: Thanks for the info, I will take a look one more time.

Comment (faloj) [reply drivers/counter/counter_ll_stm32_rtc.c:226]: Well actually as LL_RTC_EnableShadowRegBypass is called this won't be true. Even if bypass was not enabled it won't work for the STM32F2 as @gautierg-st pointed out in #62304.

Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c:226]: True, thanks for the point.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 9]
```diff
@@ -477,30 +554,21 @@ static const struct stm32_pclken rtc_clk[] = STM32_DT_INST_CLOCKS(0);
 static const struct rtc_stm32_config rtc_config = {
 	.counter_info = {
 		.max_top_value = UINT32_MAX,
-		.freq = 1,
+#ifndef CONFIG_COUNTER_RTC_STM32_SUBSECONDS
+		/* freq = 1Hz for not subsec based driver */
+		.freq = RTCCLK_FREQ / ((RTC_ASYNCPRE + 1) * (RTC_SYNCPRE + 1)),
+#else /* !CONFIG_COUNTER_RTC_STM32_SUBSECONDS */
+		.freq = RTCCLK_FREQ / (RTC_ASYNCPRE + 1),
+#endif /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS */
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:562]: I might have missed something but in BCD mode, freq is always `RTCCLK_FREQ / ((RTC_ASYNCPRE + 1) * (RTC_SYNCPRE + 1))`

Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c:562]: It depends which freq you mean :-) the subsecond register is clocked with signal after only one prescaler - the async one. So when we are using subsecond as a base, this freq is used as a reference.

File: tests/drivers/counter/counter_basic_api/testcase.yaml [hunk 10]
```diff
@@ -18,3 +18,12 @@ tests:
       - CONFIG_ZERO_LATENCY_IRQS=y
     extra_args:
       DTC_OVERLAY_FILE="boards/nrf52840dk_nrf52840.overlay;boards/nrf52840dk_nrf52840_zli.overlay"
+  drivers.counter.basic_api.stm32_subsec:
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply tests/drivers/counter/counter_basic_api/testcase.yaml:21]: What do you mean by specific test? Unfortunately, I won't be able to implement a new test at the moment.

File: drivers/counter/Kconfig.stm32_rtc [hunk 11]
```diff
@@ -44,4 +44,13 @@ config COUNTER_RTC_STM32_SAVE_VALUE_BETWEEN_RESETS
 	help
 	  Keep the counter value after each reset.
 
+config COUNTER_RTC_STM32_SUBSECONDS_BASED
```
[the remark this answers is not in the document]
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/Kconfig.stm32_rtc]: Done

File: tests/drivers/counter/counter_basic_api/testcase.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (erwango) [reply tests/drivers/counter/counter_basic_api/testcase.yaml:21]: Some way to test if `get_value_64` is not completely off.
Anyway, I mentioned this was not blocking so nvm. I think we should take care at project level that tests are added directly each time a new API comes in.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Do you want me to do anything else?

**Comment by erwango** [comment]
> Do you want me to do anything else?

Not for now, but before approving I'd like to make some tests or at least compile on few configurations.
This will also let time for others to review.

**Comment by erwango** [comment]
Here is the log:
```
*** Booting Zephyr OS build zephyr-v3.4.0-2458-gcfb5f828b708 ***
Running TESTSUITE counter_basic
===================================================================
START - test_all_channels
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing rtc@40002800
PASS - test_all_channels in 4.022 seconds
===================================================================
START - test_cancelled_alarm_does_not_expire
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing rtc@40002800
PASS - test_cancelled_alarm_does_not_expire in 7.891 seconds
===================================================================
START - test_late_alarm
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Skipped for rtc@40002800
PASS - test_late_alarm in 0.813 seconds
===================================================================
START - test_late_alarm_error
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Skipped for rtc@40002800
PASS - test_late_alarm_error in 0.813 seconds
===================================================================
START - test_multiple_alarms
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Skipped for counter
Skipped for counter
Skipped for rtc@40002800
PASS - test_multiple_alarms in 1.083 seconds
===================================================================
START - test_set_top_value_with_alarm
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Skipped for rtc@40002800
PASS - test_set_top_value_with_alarm in 1.576 seconds
===================================================================
START - test_short_relative_alarm
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing rtc@40002800
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:801: test_short_relative_alarm_instance: (i + 1 not equal to cnt)
rtc@40002800: Expected 6 callbacks, got 5
FAIL - test_short_relative_alarm in 0.753 seconds
===================================================================
START - test_single_shot_alarm_notop
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing rtc@40002800
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:403: test_single_shot_alarm_instance: (0 not equal to err)
rtc@40002800: Counter set alarm failed (err: -16)
FAIL - test_single_shot_alarm_notop in 1.221 seconds
===================================================================
START - test_single_shot_alarm_top
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Skipped for rtc@40002800
PASS - test_single_shot_alarm_top in 1.303 seconds
===================================================================
TESTSUITE counter_basic failed.
Running TESTSUITE counter_no_callback
===================================================================
START - test_set_top_value_without_alarm
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Testing counter
Skipped for rtc@40002800
PASS - test_set_top_value_without_alarm in 0.848 seconds
===================================================================
TESTSUITE counter_no_callback succeeded
------ TESTSUITE SUMMARY START ------
SUITE FAIL -  77.78% [counter_basic]: pass = 7, fail = 2, skip = 0, total = 9 duration = 19.475 seconds
- PASS - [counter_basic.test_all_channels] duration = 4.022 seconds
- PASS - [counter_basic.test_cancelled_alarm_does_not_expire] duration = 7.891 seconds
- PASS - [counter_basic.test_late_alarm] duration = 0.813 seconds
- PASS - [counter_basic.test_late_alarm_error] duration = 0.813 seconds
- PASS - [counter_basic.test_multiple_alarms] duration = 1.083 seconds
- PASS - [counter_basic.test_set_top_value_with_alarm] duration = 1.576 seconds
- FAIL - [counter_basic.test_short_relative_alarm] duration = 0.753 seconds
- FAIL - [counter_basic.test_single_shot_alarm_notop] duration = 1.221 seconds
- PASS - [counter_basic.test_single_shot_alarm_top] duration = 1.303 seconds
SUITE PASS - 100.00% [counter_no_callback]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.848 seconds
- PASS - [counter_no_callback.test_set_top_value_without_alarm] duration = 0.848 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
RunID: c68cf89ac73c546729e0323ac92016ff
PROJECT EXECUTION FAILED

```

**Comment by semihalf-niedzwiecki-dawid** [comment]
Thanks for the logs. How do you run the tests? It's executed on the `nucleo_f429zi` board, right? I've ordered the board to do it on my side and make sure everything is working fine. I should be able to reproduce the tests in a few days.

**Comment by semihalf-niedzwiecki-dawid** [comment]
I've managed to reproduce the tests on the nucleo_f429zi board. Working on the fix.

**Comment by semihalf-niedzwiecki-dawid** [comment]
I've debugged the problem.

The test failures occur when 1 tick alarm is programmed. For example  `test_short_relative_alarm` test checks the 1 tick alarm 100 times. The alarm is not triggered in ~20% cases.

I've noticed that the alarm doesn't occur when it is programmed one tick before deadline. It is a case when one tick occurs during the calculation of register values in the `set_alarm` function. IMO it still should work, but I started reading the reference manual one more time and I found the sentence `Each change of the RTC_CR register is taken into account after 1 to 2 RTCCLK clock cycles due to clock synchronization.`.

I think it means 1 tick alarm won't work for RTC with subseconds and an additional check should be added that the relative minimum tick value is 2.

Please let me know what you think.

**Comment by erwango** [comment]
> @erwango Please let me know what you think.

It seems there's a good amount of drivers dealing with such cases already. See [STM32 timer based counter driver](https://github.com/zephyrproject-rtos/zephyr/blame/main/drivers/counter/counter_ll_stm32_timer.c#L230) for instance.
Would you be able to apply the same in current case ?

**Comment by semihalf-niedzwiecki-dawid** [comment]
> > @erwango Please let me know what you think.
> 
> It seems there's a good amount of drivers dealing with such cases already. See [STM32 timer based counter driver](https://github.com/zephyrproject-rtos/zephyr/blame/main/drivers/counter/counter_ll_stm32_timer.c#L230) for instance. Would you be able to apply the same in current case ?

I had to use a bit different detection method. It works fine and passes all testes.

**Comment by erwango** [comment]
> I had to use a bit different detection method. It works fine and passes all testes.

Thanks I'll have a look once I have free time.

Just a side note, when updating a PR, please don't rebase at the same time. This prevents to easily see the diff using the `[compare](https://github.com/zephyrproject-rtos/zephyr/compare/6b87b0597c6b94fb6b4b25474f5e4c473c1ffa6d..6ce4d587861db359b7e36337b55e7f4669f70c69)` button provided by github and  we have to review the whole change again.
If rebase is really required, please proceed in 2 steps:
rebase and force push then update PR and force push (or in the order that better suit you).

If everyone would do the same, it would save a lot of time to reviewers.

**Comment by semihalf-niedzwiecki-dawid** [comment]
> Just a side note, when updating a PR, please don't rebase at the same time. This prevents to easily see the diff using the `[compare](https://github.com/zephyrproject-rtos/zephyr/compare/6b87b0597c6b94fb6b4b25474f5e4c473c1ffa6d..6ce4d587861db359b7e36337b55e7f4669f70c69)` button provided by github and we have to review the whole change again. If rebase is really required, please proceed in 2 steps: rebase and force push then update PR and force push (or in the order that better suit you).
> 
> If everyone would do the same, it would save a lot of time to reviewers.

Thanks for the tip. I just noticed how it works.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Sorry for the noise. I run the auto-formatter

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 12]
```diff
@@ -225,7 +261,12 @@ static uint32_t rtc_stm32_read(const struct device *dev)
 	ts -= T_TIME_OFFSET;
 
 	__ASSERT(sizeof(time_t) == 8, "unexpected time_t definition");
-	ticks = counter_us_to_ticks(dev, ts * USEC_PER_SEC);
+
+	ticks = ts * counter_get_frequency(dev);
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS
+	ticks += RTC_SYNCPRE - rtc_subseconds;
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:267]: A quick comment on this would be great

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 13]
```diff
@@ -180,13 +205,16 @@ static int rtc_stm32_stop(const struct device *dev)
 	return 0;
 }
 
-
-static uint32_t rtc_stm32_read(const struct device *dev)
+tick_t rtc_stm32_read(const struct device *dev)
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:208]: Non blocking but an additional cleanup commit which makes 2 variants of this functions depending on `COUNTER_NO_DATE` would be great for readability.
Interleaving both cases in the same function makes it unclear.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 14]
```diff
@@ -225,7 +261,12 @@ static uint32_t rtc_stm32_read(const struct device *dev)
 	ts -= T_TIME_OFFSET;
 
 	__ASSERT(sizeof(time_t) == 8, "unexpected time_t definition");
-	ticks = counter_us_to_ticks(dev, ts * USEC_PER_SEC);
+
+	ticks = ts * counter_get_frequency(dev);
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:265]: Since this change impacts non COUNTER_RTC_STM32_SUBSECONDS code, it would be preferable to make it in a dedicated commit with proper justification.
Same for all changes which fall in this category (you can group them per "theme")

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 15]
```diff
@@ -517,16 +619,18 @@ static const struct rtc_stm32_config rtc_config = {
...
+	.stop = rtc_stm32_stop,
+	.get_value = rtc_stm32_get_value,
+#ifdef CONFIG_COUNTER_RTC_STM32_SUBSECONDS
+	.get_value_64 = rtc_stm32_get_value_64,
+#endif /* CONFIG_COUNTER_RTC_STM32_SUBSECONDS */
+	.set_alarm = rtc_stm32_set_alarm,
+	.cancel_alarm = rtc_stm32_cancel_alarm,
+	.set_top_value = rtc_stm32_set_top_value,
+	.get_pending_int = rtc_stm32_get_pending_int,
+	.get_top_value = rtc_stm32_get_top_value,
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:633]: Non blocking: Formatting changes should be made in their own commit.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 15 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c:638]: Done.

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 13 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Done.

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 12 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c:263]: Done.

File: drivers/counter/counter_ll_stm32_rtc.c (hunk 14 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/counter/counter_ll_stm32_rtc.c]: Done.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Thanks for the comments. I should have done the requested changes by myself.

**Review by erwango — APPROVED** [review]
LGTM, but I'd like @gautierg-st to approve as well.

**Review by gautierg-st — APPROVED** [review]
LGTM
